### PR TITLE
chore: Use instance setting for self-hosted person on events

### PR DESCRIPTION
**Pull request by neilkakkar**
## Problem

Creates a toggle for enabling person on events on self-hosted instances

image

And allows us to track who's been switching it on/off.

Will write a few tests around the async migration requirements.

## Changes

👉 *Stay up-to-date with [PostHog coding conventions](https://posthog.com/docs/contribute/coding-conventions) for a smoother review.*

## How did you test this code?

Ran locally. Test that

1. Events are sent when instance settings change.
2. Simulate cloud, see the feature flag toggling works
3. Simulate self-hosted, see the instance setting works.

### CODE REVIEW

File: .github/actions/run-backend-tests/action.yml [hunk 1]
```diff
@@ -122,7 +122,7 @@ runs:
           if: ${{ inputs.ee == 'true' && inputs.group == '1' }}
           shell: bash
           run: |
-              pytest posthog -m "ee" --durations=100 --durations-min=1.0
+              PERSON_ON_EVENTS_ENABLED=true pytest posthog -m "ee" --durations=100 --durations-min=1.0
```
Comment (neilkakkar) [inline .github/actions/run-backend-tests/action.yml:125]: this should fail, just a smoke test that person_on_events tests are indeed running fine. Will revert back after CI fails

File: posthog/settings/dynamic_settings.py [hunk 2]
```diff
@@ -31,6 +31,11 @@
         "Whether to use query path using person_id, person_properties, and group_properties on events or the old query",
         str,
     ),
+    "PERSON_ON_EVENTS_ENABLED": (
```
Comment (neilkakkar) [inline posthog/settings/dynamic_settings.py:34]: will delete old instance setting in a follow up

**Review by macobo — APPROVED** [review]
:heart: Assuming test change gets reverted

File: posthog/models/team/team.py [hunk 3]
```diff
@@ -197,21 +199,22 @@ def get_effective_membership_level(self, user_id: int) -> Optional["Organization
...
-        # only when instance setting list is empty, use the feature flag
-        return posthoganalytics.feature_enabled(
-            "person-on-events-enabled",
-            str(self.uuid),
-            groups={"project": str(self.uuid)},
-            group_properties={"project": {"id": str(self.pk)}},
-            only_evaluate_locally=True,
-        )
+        # If the async migration is not complete, never enable actor on events querying.
+        if not person_on_events_ready():
```
Comment (tiina303) [inline posthog/models/team/team.py:203]: Consider excluding this check for `MULTY_TENANCY` we might be messing with the async migration on cloud and also no need to look it up every release from the async migrations table.

File: posthog/test/base.py [hunk 4]
```diff
@@ -142,10 +144,20 @@ def setUpTestData(cls):
             _setup_test_data(cls)
 
     def setUp(self):
+
+        if get_instance_setting("PERSON_ON_EVENTS_ENABLED"):
```
Comment (neilkakkar) [inline posthog/test/base.py:148]: @tiina303 not sure if this adversely affects anything for AsyncMigration tests, but open to suggestions.

The problem is: We want person on events tests to run when the instance setting is set. BUT, in all tests, the async migration isn't marked as completed, so person-on-events aren't enabled.

This is causing the smoke test to pass (when it should fail)

File: posthog/models/team/team.py (hunk 3 above)
Comment (neilkakkar) [reply posthog/models/team/team.py:203]: Makes sense

File: posthog/models/team/team.py [hunk 5]
```diff
@@ -197,21 +199,22 @@ def get_effective_membership_level(self, user_id: int) -> Optional["Organization
...
+                group_properties={"project": {"id": str(self.pk)}},
+                only_evaluate_locally=True,
+            )
+
+        # If the async migration is not complete, don't enable actor on events querying.
+        if not person_on_events_ready():
+            return False
+
+        # on self-hosted, use the instance setting
+        return get_instance_setting("PERSON_ON_EVENTS_ENABLED")
```
Comment (tiina303) [inline posthog/models/team/team.py:217]: you can now make this shorter
```suggestion
        # on self-hosted, make sure the migration is completed and use the instance setting
        return person_on_events_ready() and get_instance_setting("PERSON_ON_EVENTS_ENABLED")
```

**Comment by pauldambra** [comment]
cherry picked into 1.39.0

**Pull request merged**